sdk/core/azure-core/src/uuid.cpp

```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
+
+constexpr std::uint8_t HexToNibble(char c) // does not check for errors
 {
-  if (byte <= 9)
+  if (c >= 'a')
   {
-    return '0' + byte;
+    return 10 + (c - 'a');
   }
 
-  AZURE_ASSERT_MSG(
-      byte >= 10 && byte <= 15,
-      "It is expected, for a valid Uuid, to have byte values, where each of the two nibbles fit "
-      "into a hexadecimal character");
+  if (c >= 'A')
+  {
+    return 10 + (c - 'A');
+  }
 
-  return 'a' + (byte - 10);
+  return c - '0';
+}
+
+constexpr char NibbleToHex(std::uint8_t nibble) // does not check for errors
+{
+  if (nibble <= 9)
+  {
+    return '0' + nibble;
+  }
+
+  return 'a' + (nibble - 10);
 }
 } // namespace
 
 namespace Azure { namespace Core {
   std::string Uuid::ToString() const
   {
-    std::string s(36, '-');
+    std::string s(UuidStringLength, '-');
 
-    for (size_t i = 0, j = 0; j < s.size() && i < UuidSize; i++)
+    for (size_t bi = 0, si = 0; bi < m_uuid.size(); ++bi)
     {
-      if (i == 4 || i == 6 || i == 8 || i == 10)
+      if (IsDashIndex(si))
       {
-        j++; // Add hyphens at the appropriate places
+        ++si;
       }
 
-      uint8_t highNibble = (m_uuid[i] >> 4) & 0x0F;
-      uint8_t lowNibble = m_uuid[i] & 0x0F;
-      s[j++] = ByteToHexChar(highNibble);
-      s[j++] = ByteToHexChar(lowNibble);
+      assert((si < UuidStringLength) && (si + 1 < UuidStringLength));
+
+      const std::uint8_t b = m_uuid[bi];
+      s[si] = NibbleToHex((b >> 4) & 0x0F);
+      s[si + 1] = NibbleToHex(b & 0x0F);
+      si += 2;
     }
+
     return s;
   }
 
+  Uuid Uuid::Parse(std::string const& s)
+  {
+    bool parseError = false;
+    Uuid result;
+    if (s.size() != UuidStringLength)
+    {
+      parseError = true;
+    }
+    else
+    {
+      for (size_t si = 0, bi = 0; si < UuidStringLength; ++si)
+      {
+        const auto c = s[si];
+        if (IsDashIndex(si))
+        {
+          if (c != '-')
+          {
+            parseError = true;
+            break;
+          }
+        }
+        else
+        {
+          assert(si + 1 < UuidStringLength && bi < result.m_uuid.size());
+
+          const auto c2 = s[si + 1];
+          if (!StringExtensions::IsHexDigit(c) || !StringExtensions::IsHexDigit(c2))
+          {
+            parseError = true;
+            break;
+          }
+
+          result.m_uuid[bi] = (HexToNibble(c) << 4) | HexToNibble(c2);
+          ++si;
+          ++bi;
+        }
+      }
+    }
+
+    return parseError ? throw std::invalid_argument(
+               "Error parsing Uuid: '" + s
+               + "' is not in the '00112233-4455-6677-8899-aAbBcCdDeEfF' format.")
+                      : result;
+  }
+
   Uuid Uuid::CreateUuid()
   {
-    uint8_t uuid[UuidSize] = {};
+    Uuid result{};
+
+    // Using RngResultType and RngResultSize to highlight the places where the same type is used.
+    using RngResultType = std::uint32_t;
+    static_assert(sizeof(RngResultType) == 4, "sizeof(RngResultType) must be 4.");
```

Comment:
Tbh, introducing these two "types" makes the rest of the implementation mor complicate to read and parse.

I don't think it's worth it here. Aliasing uint32 in particular adds more confusion than the value we are trying to provide here.

Reply:
But if what we use in place of `RngResultType` is not effectively a `std::uint32_t` on some implementation, and if it's size is not `4`, it may behave incorrectly, and not be caught for a long time.
Here, we have a `static_assert` ensuring that our assumptions are correct. If it ever explodes, you need to review/update all the lines below that use `RngResultType` or `RngResultSize`. That is why they are dedicated named constants, because they express our assumptions and you don't have to review any of the lines that do not involve these two constants, if the `static_assert` explodes.

Reply:
When is it not 4? We have had this code as is for quite a while now, and have heard no issues  on any compiler/OS.

Reply:
We don't know if it has never been 4, most likely not, but now we have a `static_assert` which will fail to compile if it is not 4.